Refactor

DerekLeeCS opened the pull request:
@luster 
Can you take a look at the Abstract vs. Interface that I used?
Store is the Abstract class, and UserStoreImpl uses it.
PostStore is the Interface class, and PostStoreImpl uses it.

I'm thinking of using the Abstract class for Users, Posts, and Images.

DerekLeeCS committed:
started refactoring

Post classes should be finished

DerekLeeCS committed:
updated some code in Server

still incomplete

DerekLeeCS committed:
continued refactoring - added abstract Store class

also trimmed User class and added TrackerID interface class

luster commented on Greet-Dough/.idea/uiDesigner.xml:
```diff
@@ -0,0 +1,124 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
you should remove/add `.idea` to a `.gitignore` file in the root of the project to remove IDE files from IntelliJ

luster commented on Greet-Dough/src/main/java/store/PostStore.java:
```diff
@@ -0,0 +1,14 @@
+import model.Post;
+import utility.TrackerID;
+
+public interface PostStore extends TrackerID {
```
if for some reason each store has its own way to assign ids, it probably makes sense to allow each store to assign uuid/int/etc. i would maybe opt to keep `public interface PostStore` with no extended functionality, and then you can do something like `public class PostStoreInMemoryImpl extends TrackerId implements PostStore` since it's bound to the implementation and not so much to the interface definition

luster commented on Greet-Dough/src/main/java/store/PostStoreImpl.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Override
+    public int getFreeID() {
```
you're overriding this from the `TrackerID`. might as well just have each store maintain its own AtomicInteger/HashMap, and then you don't need the TrackerID concept necessarily

luster commented on Greet-Dough/src/main/java/utility/Store.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.HashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class Store<T> {
```
i like this abstract class - i think this encapsulates the basic concept better than the TrackerID

luster reviewed (commented):
Looks pretty good - I think you have to clean up some of the code a bit since some stuff looks a bit duplicated. the `Store` abstract class is a good thread to continue pulling on

Commit:
Move IO functions to utility class

Commit:
Rewrote User Endpoints

Tested it, works fine

Everything except put() because IDK what we are doing with that yet

Also fixed Store abstract class a bit, was missing an empty initialization for the hashMap which didnt let add/delete work at all

DerekLeeCS committed:
reworked PostStoreImpl class

DerekLeeCS committed:
wrote ImageStoreImpl class

DerekLeeCS commented:
@luster 
How should we add the subscriptions/followers for a given user?
I was thinking of having a separate SubStore to keep track of it. The other idea was to put this in the UserStoreImpl class.

Commit:
Finished post endpoints + style

We should probably start using the response variable of requests more often for sending back informative responses.

Right now im just sending back simple strings, but we can expand this to descriptive things using JSON

Commit:
added gson dependency

& sprinkled it around

Now we are returning JSON for our requests, which is more appropriate than pure 100% no cap 🔥🔥 objects🔥

luster commented:
> @luster
> How should we add the subscriptions/followers for a given user?
> I was thinking of having a separate SubStore to keep track of it. The other idea was to put this in the UserStoreImpl class.

Could have a SubStore with a List of user ids for who they’re following, keyed by user id. It could also be part of the UserStore, but you’ll wanna be careful of too many features creeping in to the User concept. Play around with it a bit!

doanhub committed:
Create Likes.java

doanhub committed:
Create Comments.java

doanhub committed:
Fix code style

DerekLeeCS committed:
refactored Store class and wrote Relation class

Store inherits from StorageRetrieval
Relation also inherits from StorageRetrieval
Planning to use Relation for SubStore and FollowStore

DerekLeeCS committed:
Merge branch 'refactor' of https://github.com/cooperece366s21/Lee-Ko into refactor

DerekLeeCS committed:
updated code style

DerekLeeCS committed:
updated StorageRetrieval to abstract class

DerekLeeCS commented:
@luster 
Thoughts on return values for success/failure in functions?
E.g. 
Boolean: 1 = success, 0 = failure
Int: 0 = success, -1 = failure for reason #1, -2 = failure for reason #2, etc.

DerekLeeCS committed:
wrote FollowStoreImpl and reorganized class files

luster commented:
> @luster
> Thoughts on return values for success/failure in functions?
> E.g.
> Boolean: 1 = success, 0 = failure
> Int: 0 = success, -1 = failure for reason 1, -2 = failure for reason 2, etc.

I usually prefer an object, or Optional<T>. Can have some kind of Result<T> with the optional value as well as some kind of optional error enum. Would prefer that over Primitive Integer codes for this kind of use case

DerekLeeCS commented:
> > @luster
> > Thoughts on return values for success/failure in functions?
> > E.g.
> > Boolean: 1 = success, 0 = failure
> > Int: 0 = success, -1 = failure for reason 1, -2 = failure for reason 2, etc.
> 
> I usually prefer an object, or Optional. Can have some kind of Result with the optional value as well as some kind of optional error enum. Would prefer that over Primitive Integer codes for this kind of use case

So I should create a Result class, and use a custom enum for each method?

doanhub committed:
Updated Image, Likes

Stevenree committed:
Image Constructor/getter and setter

ThodoriKapouranis commented:
@luster 

We are considering different ways to implement the image class for posts, and would like your input. 
**We plan on having some sort of file hosting/storage for embedded files in posts

1. Should we be treating images similar to users and posts, where the images related to a post are fetched using a ImagePostStore? In terms of frontend, the user will only be able to place these files that are uploaded for the current post.

2. Also I am thinking about saving post content as the HTML that would render that post. This means that it'll have references to other files (ex: with <img> tags) but not nested within the post class itself. Is this potentially a bad idea? All referenced files are ones uploaded by (1.)

doanhub committed:
Updated Image, Likes

doanhub committed:
Update Comment, CommentStoreImpl

luster commented:
> > > @luster
> > > Thoughts on return values for success/failure in functions?
> > > E.g.
> > > Boolean: 1 = success, 0 = failure
> > > Int: 0 = success, -1 = failure for reason 1, -2 = failure for reason 2, etc.
> > 
> > 
> > I usually prefer an object, or Optional. Can have some kind of Result with the optional value as well as some kind of optional error enum. Would prefer that over Primitive Integer codes for this kind of use case
> 
> So I should create a Result class, and use a custom enum for each method?

You don't have to be that robust about it. Typically you can get away with using `Optional<T>`. If the return type is empty, that's usually an error. For example, if you look up a user in a db, it can be (a) found, (b) not found, or (c) some kind of error while looking up. In the error case, you likely want to throw that error up through the code and have the handler return an HTTP 500 INTERNAL SERVER ERROR. If (a), HTTP 200 with the found object returned, and if (b) HTTP 404 with the object not found. 

Typically, you'll want to address error handling like that. Let the higher level http handler take care of error types, and let the lower level optimize for the "happy path." The more custom error codes you have, the more logic you have to write around it. Hope this makes sense!

luster commented:
> @luster
> 
> We are considering different ways to implement the image class for posts, and would like your input.
> **We plan on having some sort of file hosting/storage for embedded files in posts
> 
> Should we be treating images similar to users and posts, where the images related to a post are fetched using a ImagePostStore? In terms of frontend, the user will only be able to place these files that are uploaded for the current post.
> 
> Also I am thinking about saving post content as the HTML that would render that post. This means that it'll have references to other files (ex: with  tags) but not nested within the post class itself. Is this potentially a bad idea? All referenced files are ones uploaded by (1.)

Hmm, I don't know that I fully understand the questions. If you're uploading images and need those referenced by other user generated content, you would typically save the image to some kind of blob storage (local filesystem for testing, Google Cloud Storage, Amazon S3, etc), and then you would save the link/url/path to that file as a column/attribute in your database. For instance, user uploads cat.png, that transfers the file to somewhere like gs://lee-ko/images/cat-d1e2c3.png, and then you save that path to the post table, i.e. post id=1, image url=gs://..., text='yo its a cat', etc

DerekLeeCS committed:
cleaned up code

Stevenree commented:
@luster 
I am asking regarding the likes and comment that we are implementing, I recall you asking to separate the implementation and the actual storage but I was not sure if it can be edited outside of the class. 
Second thing is the comment:
we are currently using a system where we are mapping the time of comment to a list of {user, content} but we are not sure if that is the best way. I am wondering if there are some better ways of storing the comments/user.

DerekLeeCS commented:
@luster 
Can you take a look at the relation package? Thoughts on how it's structured/how to improve?

DerekLeeCS committed:
updated return types of Relation package to void

todo: error checking at a higher level

DerekLeeCS committed:
created handler class

Commit:
Move server functionality to handler class

- replaced endpoint strings with actual url

- moved nearly all server functionality to handler class

- Should we return responses with object as JSON in body, or object itself? @luster

😊

luster commented:
Ack that I see both of your comments @Stevenree @DerekLeeCS. I'll get back to you tomorrow on this!

thodorii committed:
Add getFeed() for PostStoreImplem

luster commented:
> @luster
> I am asking regarding the likes and comment that we are implementing, I recall you asking to separate the implementation and the actual storage but I was not sure if it can be edited outside of the class.
> Second thing is the comment:
> we are currently using a system where we are mapping the time of comment to a list of {user, content} but we are not sure if that is the best way. I am wondering if there are some better ways of storing the comments/user.

on the first point: model class should hold the values only. `service` classes should control application-specific logic which will (1) construct the model classes you're looking to save/update/etc in the `store` classes, and (2) get/update/delete/etc from the store classes

on the second point: i'm not sure about comments specifically. if you leave a comment on GitHub at the specific line of code, it's easier for me to provide feedback for you. but, typically for a "comment" on a "post," you'd want some way to map comments to the post, and you want to know the authors of certain comments. feel free to use ArrayList/HashMap in order to track these sorts of things. It's ok to iterate over all entries in a map just for testing purposes if, for instance, you have to grab all comments or posts from a specific author.

thodorii committed:
Added subscribe(), unsubscribe()

Also updated the feed so that correct permissions are needed

Theres a lot of conditional code to check for errors, a little repetetive, maybe we can wrap a lot in functions or something idk

luster commented on Greet-Dough/src/main/java/store/relation/SubStoreImpl.java:
```diff
@@ -0,0 +1,31 @@
+
+// Store subscriptions as ArrayList or HashMap?
+// A single user's subscriptions probably won't be so large that a linear search will take very long
+public class SubStoreImpl extends Relation {
```
this looks good - i took a look in IntelliJ at the object hierarchy. 

my recommendation would be to do something like `public class SubStoreImpl extends Relation implements SubStore` and abstract the public methods in your Store classes. this way, when you use a database instead of hashmaps/lists/etc, you can just re-implement the top level interface.

intellij has an "extract interface" refactoring you can use so that you don't have to rewrite everything. in any case, it should be straightforward to do manually as well!

then, in your Server, where you're instantiating the store, you can actually do `SubStore subStore = new SubStoreImpl()`, and the classes that depend on it can also depend on `SubStore` instead of `SubStoreImpl`. down the road, super easy to just change the implementation to `SubStorePostgres` in the Server main method instead of everywhere in the code

hope that makes sense!

luster replied:
cc @DerekLeeCS

thodorii committed:
Add handler's follow() unfollow() and endpoints

DerekLeeCS committed:
deleted main.java

DerekLeeCS committed:
Merge branch 'refactor' of https://github.com/cooperece366s21/Lee-Ko into refactor

doanhub committed:
Update Comment, CommentStoreImpl

doanhub committed:
Update Comment, Likes Implementations

Stevenree commented:
@luster 
Can u take a look at the likeimpl and commentImpl that we did. I asked about how to store the comment before and this is what we came up with at the moment. Thank you!

DerekLeeCS committed:
updated Likes, LikeStoreImpl, Handler class

- moved attemptLike (renamed to likePost) logic to Handler class
- updated code style

luster commented:
> @luster
> Can u take a look at the likeimpl and commentImpl that we did. I asked about how to store the comment before and this is what we came up with at the moment. Thank you!

implementation seems fine - abstract store class which has the id generation and hash maps. you can keep the underlying store interfaces when you switch to a real database and just ditch the abstracts.

one note on the comment interface: why does adding a comment require both `String commentContent` and `Comment newComment`? seems like it could be abstracted better. to be honest, could you just have `void addComment(Comment comment)`, with the `Comment` object containing both the userId of who is making the comment as well as the content itself?

otherwise, nice job extracting some commonalities to make implementing the stores easier up front!

Stevenree commented:
@luster 
for the class that we have in comment rn, we just have a storage of ids and hash that stores the comments. 
do u think it would be better if we create a subclass from comment that has a comment id, the content and the userid of who commented it, then we can just map the time commented to the comment id which can pull up the content easier?

luster commented:
I'm not sure I'm fully getting the distinction. From what I saw, things look fine for now. You'll likely want nested comments to also be comments, but with an `Optional<Integer> parentCommentId` or something
